File: garbler.py
from dataclasses import dataclass
from random import shuffle
from typing import Optional

from cryptography.fernet import Fernet

from utils import GateType, bytes_xor, encrypt, generate_keys, topo_sort_wires


@dataclass
class GarbledGate:
    operation: GateType
    table: dict[tuple[int, Optional[int]], bytes]  # lhs, rhs -> output
    input_keys: dict[str, tuple[bytes, bytes]]  # wire_id -> (0_label, 1_label)
    output_keys: tuple[bytes, bytes]


class Garbler:
    def __init__(
        self,
        circuit: dict[str, Optional[tuple[GateType, str, Optional[str]]]],
        ins: dict[str, list[int]],
        outs: dict[str, list[int]],
    ):
        self.__delta = Fernet.generate_key()
        self.circuit = circuit
        self.wire_to_keys: dict[str, tuple[bytes, bytes]] = {}
        # gate_id -> garbled_gate
        self.garbled_gates: dict[str, GarbledGate] = {}

        self.circuit = topo_sort_wires(circuit, ins, outs)

        for wire_name, wire_id in ins.items():
            for i in wire_id:
                self.wire_to_keys[f"{wire_name}_{i}"] = generate_keys(self.__delta)

        for wire_name, wire_id in outs.items():
            for i in wire_id:
                self.wire_to_keys[f"{wire_name}_{i}"] = generate_keys(self.__delta)

        for gate_wire, gate in self.circuit.items():
            if gate is not None:
                print(
                    f"Garbling gate {gate[0]} for wire {gate_wire} with inputs {gate[1:]}"
                )
                garbled = self.garble_gate(
                    gate[0],
                    {wire: self.wire_to_keys[wire] for wire in gate[1:]},
                    self.wire_to_keys.get(gate_wire) if gate_wire in outs else None,
                )
                self.wire_to_keys[gate_wire] = (
                    garbled.output_keys[0],
                    garbled.output_keys[1],
                )
                self.garbled_gates[gate_wire] = garbled

    def garble_gate(
        self,
        gate_op: GateType,
        wire_to_keys: dict[str, tuple[bytes, bytes]],
        output: Optional[tuple[bytes, bytes]],
    ) -> GarbledGate:
        if gate_op == GateType.NOT:
            assert len(wire_to_keys) == 1, "NOT gate only has one input"
        else:
            assert len(wire_to_keys) == 2, "Gate expects have 2 inputs"

        # a_val, b_val -> output
        garbled_table: dict[tuple[int, Optional[int]], bytes] = {}

        wire_ids = list(wire_to_keys.keys())
        in_key_a_0, in_key_a_1 = wire_to_keys[wire_ids[0]]
        in_key_b_0, in_key_b_1 = (
            wire_to_keys[wire_ids[1]] if len(wire_ids) > 1 else (None, None)
        )

        if gate_op == GateType.XOR:
            assert in_key_b_0 is not None, "XOR gate must have 2 inputs"
            output_0_label = bytes_xor(in_key_a_0, in_key_b_0)
            output_1_label = bytes_xor(output_0_label, self.__delta)

            return GarbledGate(
                gate_op, {}, wire_to_keys, (output_0_label, output_1_label)
            )

        if output is None:
            output_0_label, output_1_label = generate_keys(self.__delta)
        else:
            output_0_label, output_1_label = output

        for a_val in (0, 1):
            for b_val in (0, 1) if gate_op != GateType.NOT else [None]:
                if gate_op == GateType.NOT:
                    in_key_a = wire_to_keys[wire_ids[0]][a_val]
                    out_val = not a_val
                    out_bytes_val = output_1_label if out_val else output_0_label
                    garbled_table[(a_val, None)] = encrypt(in_key_a, out_bytes_val)
                else:
                    in_key_a = in_key_a_0 if a_val == 0 else in_key_a_1
                    in_key_b = in_key_b_0 if b_val == 0 else in_key_b_1

                    switch = {
                        GateType.AND: lambda a_val, b_val: a_val and b_val,
                        GateType.OR: lambda a_val, b_val: a_val or b_val,
                        GateType.NOR: lambda a_val, b_val: not (a_val or b_val),
                        GateType.ORNOT: lambda a_val, b_val: a_val or (not b_val),
                        GateType.NAND: lambda a_val, b_val: not (a_val and b_val),
                        GateType.ANDNOT: lambda a_val, b_val: a_val and (not b_val),
                        GateType.XNOR: lambda a_val, b_val: not (a_val != b_val),
                    }
                    out_op = switch[gate_op]
                    out_val = out_op(a_val, b_val)
                    out_bytes_val = output_1_label if out_val else output_0_label
                    garbled_table[(a_val, b_val)] = encrypt(
                        in_key_a, encrypt(in_key_b, out_bytes_val)
                    )

        items = list(garbled_table.items())
        shuffle(items)
        shuffled_table = dict(items)
        return GarbledGate(
            gate_op, shuffled_table, wire_to_keys, (output_0_label, output_1_label)
        )

File: src/garbler.rs
use std::collections::BTreeMap;
use std::option::Option;

use base64::{engine::general_purpose::URL_SAFE, Engine as _};

use crate::utils::{
    bytes_xor, encrypt, generate_encryption_key, generate_keys, topo_sort_wires, Circuit, GateType,
};
use rand::seq::SliceRandom;
use rand::thread_rng;

#[derive(Debug, Clone)]
pub struct GarbledGate {
    pub operation: GateType,
    pub table: Vec<String>,
    pub input_wire_ids: Vec<String>,
    pub output_keys: (String, String),
}

#[derive(Debug)]
pub struct Garbler {
    delta: String,
    circuit: Circuit,
    wire_to_keys: BTreeMap<String, (String, String)>,
    garbled_gates: BTreeMap<String, GarbledGate>,
}

impl Garbler {
    pub fn new(
        delta: String,
        circuit: Circuit,
        ins: BTreeMap<String, Vec<String>>,
        outs: BTreeMap<String, Vec<String>>,
    ) -> Self {
        let mut wire_to_keys: BTreeMap<String, (String, String)> = BTreeMap::new();

        for (_wire_name, wire_id) in ins.iter() {
            for i in wire_id {
                wire_to_keys.insert(i.to_string(), generate_keys(&delta));
            }
        }

        for (_wire_name, wire_id) in outs.iter() {
            for i in wire_id {
                wire_to_keys.insert(i.to_string(), generate_keys(&delta));
            }
        }

        Self {
            delta,
            circuit,
            wire_to_keys,
            garbled_gates: BTreeMap::new(),
        }
    }

    pub fn build(
        &mut self,
    ) -> (
        BTreeMap<String, (String, String)>,
        BTreeMap<String, GarbledGate>,
    ) {
        let sorted_wires = topo_sort_wires(&self.circuit);
        for wire in sorted_wires {
            let (gate_type, gate_inputs) = self.circuit.get(&wire).unwrap();
            if gate_type == &GateType::INPUT {
                continue;
            }

            let gate_wire_to_keys_cloned: BTreeMap<String, (String, String)> = gate_inputs
                .iter()
                .map(|input_wire| {
                    let key_pair = self.wire_to_keys.get(input_wire).unwrap().clone();
                    (input_wire.clone(), key_pair)
                })
                .collect();

            let gate_wire_to_keys: BTreeMap<&String, &(String, String)> = gate_wire_to_keys_cloned
                .iter()
                .map(|(k, v)| (k, v))
                .collect();

            let output_wire_keys = self.wire_to_keys.get(&wire).cloned();
            let gg = self.garble_gate(
                *gate_type,
                gate_wire_to_keys,
                output_wire_keys,
                gate_inputs.to_vec(),
            );

            self.wire_to_keys
                .insert(wire.clone(), gg.output_keys.clone());
            self.garbled_gates.insert(wire, gg);
        }
        (self.wire_to_keys.clone(), self.garbled_gates.clone())
    }

    fn switch_gate(&self, gate_type: GateType, a_val: bool, b_val: bool) -> bool {
        match gate_type {
            GateType::AND => a_val & b_val,
            GateType::OR => a_val | b_val,
            GateType::NOR => !(a_val | b_val),
            GateType::ORNOT => a_val | !b_val,
            GateType::NAND => !(a_val & b_val),
            GateType::ANDNOT => a_val & !b_val,
            GateType::XNOR => a_val == b_val,
            GateType::XOR => a_val ^ b_val,
            GateType::NOT => !a_val,
            GateType::CONST_0 => false,
            GateType::CONST_1 => true,
            GateType::CONST => a_val,
            GateType::INPUT => a_val,
        }
    }

    pub fn garble_gate(
        &mut self,
        gate_op: GateType,
        gate_wire_to_keys: BTreeMap<&String, &(String, String)>,
        output: Option<(String, String)>,
        gate_input_names: Vec<String>,
    ) -> GarbledGate {
        assert!(
            (gate_op == GateType::NOT
                || gate_op == GateType::CONST_0
                || gate_op == GateType::CONST_1
                || gate_op == GateType::CONST)
                && gate_wire_to_keys.len() == 1
                || gate_wire_to_keys.len() == 2,
            "Invalid number of inputs for gate type"
        );

        let (in_keys_a, in_keys_b) = (
            &gate_wire_to_keys[&gate_input_names[0]],
            if gate_input_names.len() > 1 {
                gate_wire_to_keys.get(&gate_input_names[1])
            } else {
                None
            },
        );

        if gate_op == GateType::XOR {
            let safe_in_keys_a = URL_SAFE.decode(&in_keys_a.0).unwrap();
            let safe_in_keys_b = URL_SAFE.decode(&in_keys_b.unwrap().0).unwrap();
            let output_0_bytes = bytes_xor(&safe_in_keys_a, &&safe_in_keys_b);

            let safe_delta = URL_SAFE.decode(&self.delta).unwrap();
            let output_1_bytes = bytes_xor(&output_0_bytes, &safe_delta);

            let base64_output_0 = URL_SAFE.encode(&output_0_bytes);
            let base64_output_1 = URL_SAFE.encode(&output_1_bytes);

            return GarbledGate {
                operation: gate_op,
                table: vec![], // free xor gate, no table
                input_wire_ids: gate_input_names,
                output_keys: (base64_output_0, base64_output_1),
            };
        }

        let output_labels = output.unwrap_or_else(|| generate_keys(&self.delta));
        let mut garbled_table: Vec<String> = vec![];

        for (a_val, a_key) in [(0, &in_keys_a.0), (1, &in_keys_a.1)] {
            if let Some(in_keys_b) = in_keys_b {
                for (b_val, b_key) in [(0, &in_keys_b.0), (1, &in_keys_b.1)] {
                    let out_val = self.switch_gate(gate_op, a_val == 1, b_val == 1);
                    let out_bytes_val = if out_val {
                        &output_labels.1
                    } else {
                        &output_labels.0
                    };

                    let key = generate_encryption_key(&[a_key.as_bytes(), b_key.as_bytes()]);
                    let encoded_key = URL_SAFE.encode(&key);

                    garbled_table.push(encrypt(&encoded_key, out_bytes_val.as_bytes().to_vec()));
                }
            } else {
                let out_val = if a_val == 1 { 0 } else { 1 };
                let out_bytes_val = if out_val == 1 {
                    &output_labels.1
                } else {
                    &output_labels.0
                };

                let key = generate_encryption_key(&[a_key.as_bytes()]);
                let encoded_key = URL_SAFE.encode(&key);

                garbled_table.push(encrypt(&encoded_key, out_bytes_val.as_bytes().to_vec()));
            }
        }

        garbled_table.shuffle(&mut thread_rng());

        GarbledGate {
            operation: gate_op,
            table: garbled_table,
            input_wire_ids: gate_input_names,
            output_keys: (output_labels.0, output_labels.1),
        }
    }
}
